File: app/src/test/java/com/sample/android/elm/login/presenter/LoginPresenterTest.java
package com.sample.android.elm.login.presenter;

import com.sample.android.elm.data.IApiService;
import com.sample.android.elm.login.view.ILoginView;
import io.reactivex.schedulers.Schedulers;
import junit.framework.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;

public class LoginPresenterTest {

    LoginPresenter presenter;
    ILoginView view;
    IApiService loginService;

    @Before
    public void setUp() {
        view = Mockito.mock(ILoginView.class);
        loginService = Mockito.mock(IApiService.class);
        presenter = new LoginPresenter(view, loginService, Schedulers.trampoline());
    }

    @Test
    public void validateLogin() {
        Assert.assertEquals(false, presenter.validateLogin(null));
        Assert.assertEquals(false, presenter.validateLogin(""));
        Assert.assertEquals(false, presenter.validateLogin("123"));
        Assert.assertEquals(true, presenter.validateLogin("1234"));
        Assert.assertEquals(true, presenter.validateLogin("1234567891234"));
        Assert.assertEquals(false, presenter.validateLogin("1234567891234567"));
    }

    @Test
    public void validatePass() {
        Assert.assertEquals(false, presenter.validatePass(null));
        Assert.assertEquals(false, presenter.validatePass(""));
        Assert.assertEquals(false, presenter.validatePass("1234"));
        Assert.assertEquals(true, presenter.validatePass("12345"));
        Assert.assertEquals(true, presenter.validatePass("1234567891234"));
    }
}

File: app/src/test/java/com/sample/android/elm/login/presenter/LoginPresenterTest.kt
package com.sample.android.elm.login.presenter

import com.sample.android.elm.Init
import com.sample.android.elm.None
import com.sample.android.elm.Program
import com.sample.android.elm.data.IApiService
import com.sample.android.elm.data.IAppPrefs
import com.sample.android.elm.login.view.ILoginView
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import org.hamcrest.CoreMatchers.instanceOf
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.Mockito.*

class LoginPresenterTest {

    lateinit var presenter: LoginPresenter
    lateinit var view: ILoginView
    lateinit var loginService: IApiService
    lateinit var prefs: IAppPrefs
    lateinit var program: Program

    @Before
    fun setUp() {
        view = mock(ILoginView::class.java)
        loginService = mock(IApiService::class.java)
        prefs = mock(IAppPrefs::class.java)
        program = Program(Schedulers.trampoline())
        presenter = LoginPresenter(view, program, prefs, loginService, Schedulers.trampoline())
    }

    @Test
    fun initWithSavedLogin() {
        //login screen init and look for saved credentials in preferences
        var initState = LoginPresenter.LoginState()
        //update
        val (searchForLoginState, searchForLoginCmd) = presenter.update(Init(), initState)

        assertEquals(initState.copy(isLoading = true), searchForLoginState)
        assertThat(searchForLoginCmd, instanceOf(LoginPresenter.GetSavedUserCmd::class.java))

        //render
        presenter.render(searchForLoginState)
        verify(view).showProgress()
        verify(view).disableLoginBtn()
        verify(view).hideLoginError()
        verify(view).hidePasswordError()
        verify(view).hideError()
        verifyNoMoreInteractions(view)

        Mockito.`when`(prefs.getUserSavedCredentials()).thenReturn(Single.just(Pair("login", "password")))
        //call
        val loadedCredentialsMsg = presenter.call(searchForLoginCmd)

        //credentials loaded and start auth http call
        //update
        val (startAuthState, startAuthCmd) = presenter.update(loadedCredentialsMsg.blockingGet(), searchForLoginState)
        assertEquals((searchForLoginState as LoginPresenter.LoginState).copy(login = "login", pass = "password"), startAuthState)
        assertThat(startAuthCmd, instanceOf(LoginPresenter.LoginCmd::class.java))
        assertEquals("login", (startAuthCmd as LoginPresenter.LoginCmd).login)
        assertEquals("password", startAuthCmd.pass)

        Mockito.reset(view)
        //render
        presenter.render(startAuthState)
        verify(view).showProgress()
        verify(view).disableLoginBtn()
        verify(view).hideLoginError()
        verify(view).hidePasswordError()
        verify(view).hideError()
        verifyNoMoreInteractions(view)

        Mockito.`when`(loginService.login("login", "password")).thenReturn(Single.just(true))
        //call
        val authOkMsg = presenter.call(startAuthCmd)

        //auth OK, go to main screen
        //update
        val (loggedState, noneCmd) = presenter.update(authOkMsg.blockingGet(), startAuthState)
        assertThat(noneCmd, instanceOf(None::class.java))

        Mockito.reset(view)
        //render
        presenter.render(loggedState)
        verify(view).goToMainScreen()
        verify(view).hideKeyboard()
        verifyNoMoreInteractions(view)

    }


}
